dll/shellext/cabview/folder.cpp

```diff
@@ -0,0 +1,837 @@
+                for (UINT i = 0; i < data.pCIDA->cidl && hr == S_FALSE; ++i)
+                {
+                    C_ASSERT(FLATFOLDER);
+                    LPCITEMIDLIST pidlChild = ILFindLastID(HIDA_GetPIDLItem(data.pCIDA, i));
+                    CABITEM *haystack = CABITEM::Validate(pidlChild);
+                    if (!haystack && FAILED_UNEXPECTEDLY(hr = E_FAIL))
+                        break;
+                    if (!lstrcmpiW(needle->Path, haystack->Path))
+                    {
+                        if (data.pPD)
+                            data.pPD->SetLine(1, needle->Path, TRUE, NULL);
+                        hr = S_OK; // Found it in the list of files to extract
+                    }
+                }
+                SHFree(needle);
+            }
+            if (data.pPD)
+                data.pPD->SetProgress(data.completed++, data.cabfiles);
+            return hr;
+        }
+
+        case ECM_PREPAREPATH:
+        {
+            UINT flags = SHPPFW_DIRCREATE | SHPPFW_IGNOREFILENAME;
+            return SHPathPrepareForWriteW(GetUiOwner(data), NULL, ecd.Path, flags);
+        }
+
+        case ECM_ERROR:
+        {
+            return ErrorBox(GetUiOwner(data), ecd.hr);
+        }
+    }
+    return E_NOTIMPL;
+}
+
+static void Free(EXTRACTFILESDATA &data)
+{
+    if (data.pPD)
+    {
+        data.pPD->StopProgressDialog();
+        data.pPD->Release();
+    }
+    if (data.pCIDA)
+    {
+        GlobalUnlock(data.pCIDA);
+        ReleaseStgMedium(&data.cidamedium);
+    }
+    IUnknown_Set((IUnknown**)&data.pDO, NULL);
+    IUnknown_Set((IUnknown**)&data.pMarshalDO, NULL);
+    IUnknown_Set((IUnknown**)&data.pLifetimeCF, NULL);
+    SHFree(&data);
+}
+
+static DWORD CALLBACK ExtractFilesThread(LPVOID pParam)
+{
+    EXTRACTFILESDATA &data = *(EXTRACTFILESDATA*)pParam;
+    HRESULT hr = S_OK;
+    if (SUCCEEDED(SHCoCreateInstance(NULL, &CLSID_ProgressDialog, NULL, IID_PPV_ARG(IProgressDialog, &data.pPD))))
+    {
+        // TODO: IActionProgress SPACTION_COPYING
+        if (SUCCEEDED(data.pPD->StartProgressDialog(data.hWndOwner, NULL, PROGDLG_NOTIME, NULL)))
+        {
+            data.pPD->SetTitle(data.cab);
+            data.pPD->SetLine(2, data.path, TRUE, NULL);
+            data.pPD->SetAnimation(GetModuleHandleW(L"SHELL32"), 161);
+            data.pPD->SetProgress(0, 0);
+        }
+    }
+    if (data.pMarshalDO)
+    {
+        hr = CoGetInterfaceAndReleaseStream(data.pMarshalDO, IID_PPV_ARG(IDataObject, &data.pDO));
+        data.pMarshalDO = NULL;
+        if (SUCCEEDED(hr))
+        {
+            CLIPFORMAT cfhida = (CLIPFORMAT)RegisterClipboardFormatW(CFSTR_SHELLIDLISTW);
+            FORMATETC fmt = { cfhida, NULL, DVASPECT_CONTENT, -1, TYMED_HGLOBAL };
+            if (SUCCEEDED(hr = data.pDO->GetData(&fmt, &data.cidamedium)))
+            {
+                if ((data.pCIDA = (CIDA*)GlobalLock(data.cidamedium.hGlobal)) == NULL)
+                    hr = E_OUTOFMEMORY;
+            }
```

Comment:
We have `CDataObjectHIDA` to do this (from `shellutils.h`)

Reply:
CDataObjectHIDA has a constructor that takes an IDataObject pointer but here IDataObject is optional and I don't know if I have a pointer until well into the function. The best I can do is to use the recently added CDataObjectHIDA static helper functions.